### PR TITLE
Skip elements that are not of mesh dim 

### LINKED ISSUE #7335: MooseMesh::detectPairedSidesets assumes that all elements in the mesh are of the same dimension
### Description of the enhancement or error report
In a mixed dimensional mesh adding periodic boundary conditions can fail, because `MooseMesh::detectPairedSidesets()` uses `getMesh().mesh_dimension()` when calculating face normals. It loops over all elements in the mesh. If a 2D mesh contains 1D domains (for Mortar for example) the 1D elements will not have a 2D element type, causing libmesh to error out.

### Rationale for the enhancement or information for reproducing the error
Fix PBC for mixed dimensional meshes.

### Identified impact
Fixes two example problems.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #7335

### CODE REVIEW

**Comment by friedmud** [comment]
Hmmmm - I'm not really for this.  What's the idea here?

I could definitely see where we have some 2D/3D stuff going on and it could be set up so that both 2D and 3D need periodic BCs: think of a 3D box where you want to solve a set of 2D equations on the top and there are periodic BC's all around the outside...

**Comment by friedmud** [comment]
I would prefer a proper fix that matches up like-dimensioned elements...

**Comment by dschwen** [comment]
Ok, I agree, I would open a separate issue for that though. We have simulations _right now_ that can be made working with this two line fix. The current code is flat out broken and it would be great to get this fix in first.

**Comment by friedmud** [comment]
Hmmm... the reality though is that we'll never come back to this.  Meanwhile... you have your eyes on the code right now and could probably come up with a proper fix.

I'll let @permcody decide what to do 😈

**Comment by dschwen** [comment]
Are normals for 0D faces on 1D elements properly defined?

**Comment by friedmud** [comment]
Yep.  They point "outward" from the 1D domain.

**Comment by dschwen** [comment]
well, so for 1D elements embedded in a 2D or 3D domain they point in the direction of the element outwards at its ends?

**Comment by friedmud** [comment]
Yep

**Comment by dschwen** [comment]
Ok, I have a fix in mind.

**Comment by dschwen** [comment]
The latest commit will find boundary pairs within the same dimensions.

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -1281,6 +1288,10 @@ MooseMesh::detectPairedSidesets()
   {
     Elem * elem = *el;
 
+    // dimension of the current element and its normals
+    unsigned int i = elem->dim() - 1;
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: Maybe call this variable something more useful... like `side_dim`?  I actually had trouble figuring out what was going on down below...

Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: Might as well also change the loops above and below to match the variable name...

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -1299,38 +1310,41 @@ MooseMesh::detectPairedSidesets()
 
         // x-direction faces
         if (normals[0].absolute_fuzzy_equals(minus_x))
-          minus_x_ids.insert(face_ids.begin(), face_ids.end());
+          minus_x_ids[i].insert(face_ids.begin(), face_ids.end());
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: What does this do if there are no ids?  Can we short-circuit this whole logic by checking that first?

File: framework/src/mesh/MooseMesh.C [hunk 3]
```diff
@@ -1299,38 +1310,41 @@ MooseMesh::detectPairedSidesets()
...
-
-  if (minus_z_ids.size() == 1 && plus_z_ids.size() == 1)
-    _paired_boundary.push_back(std::make_pair(*(minus_z_ids.begin()),
-                                              *(plus_z_ids.begin())));
+  for (unsigned i = 0; i < dim; ++i)
+  {
+    // If unique pairings were found, fill up the _paired_boundary data
+    // structure with that information.
+    if (minus_x_ids[i].size() == 1 && plus_x_ids[i].size() == 1)
+      _paired_boundary.push_back(std::make_pair(*(minus_x_ids[i].begin()),
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: Use `emplace_back()`?

File: framework/src/mesh/MooseMesh.C [hunk 4]
```diff
@@ -1260,16 +1256,27 @@ MooseMesh::detectPairedSidesets()
...
+  std::vector<UniquePtr<FEBase>> fe_faces(dim);
+  std::vector<UniquePtr<QGauss>> qfaces(dim);
+  for (unsigned i = 0; i < dim; ++i)
+  {
+    // Face is assumed to be flat, therefore normal is assumed to be
+    // constant over the face, therefore only compute it at 1 qp.
+    qfaces[i] = UniquePtr<QGauss>(new QGauss(i, CONSTANT));
+
+    // A first-order Lagrange FE for the face.
+    fe_faces[i] = UniquePtr<FEBase>(FEBase::build(i + 1, FEType(FIRST, LAGRANGE)));
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: What's up with the creation of a `UniquePtr` here?  Isn't that already the return value of `FEBase::build()`?

File: framework/src/mesh/MooseMesh.C (hunk 2 above)
Comment (dschwen) [inline framework/src/mesh/MooseMesh.C]: Seems like premature optimization to me :-)

File: framework/src/mesh/MooseMesh.C (hunk 4 above)
Comment (dschwen) [inline framework/src/mesh/MooseMesh.C]: D'Oh... yeah, I'll change that.

File: framework/src/mesh/MooseMesh.C [hunk 5]
```diff
@@ -1299,38 +1311,41 @@ MooseMesh::detectPairedSidesets()
```
Comment (friedmud) [inline framework/src/mesh/MooseMesh.C]: How about a short-circuit here in the case of no boundary ids?

Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: This routine only gets called by the periodic bc code. No need to over optimize.

**Comment by permcody** [comment]
👍

**Pull request merged**